### PR TITLE
$translate operation implementation for remote terminology

### LINKED ISSUE #3442: Provide RemoteTerminologyServiceValidationSupport Implementation For Operation $translate
The purpose of this Issue is to enhance the current support for the $translate Operation so it can be provided by an instance of RemoteTerminologyServiceValidationSupport. It currently returns NULL instead of providing an implementation.

### PR DESCRIPTION
**Pull request by seanmikalson-smilecdr**
$translate operation implementation for remote terminology

fixes #3442

### CODE REVIEW

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/TermConceptMappingSvcImpl.java [hunk 1]
```diff
@@ -120,14 +122,27 @@ public FhirContext getFhirContext() {
...
+		for (IBaseCoding aCoding : theRequest.getCodings()) {
+			sourceCodeableConcept
+				.addCoding()
+				.setSystem(aCoding.getSystem())
+				.setCode(aCoding.getCode())
+				.setVersion(((Coding) aCoding).getVersion());
+		}
 
 		TranslationRequest request = new TranslationRequest();
 		request.setCodeableConcept(sourceCodeableConcept);
```
Comment (fil512) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/TermConceptMappingSvcImpl.java:134]: suggestion: move this to a static method:
```
TranslationRequest request = TranslationRequest.fromTranslateCodeRequest(theRequest);
```

File: hapi-fhir-validation/src/main/java/org/hl7/fhir/common/hapi/validation/support/RemoteTerminologyServiceValidationSupport.java [hunk 2]
```diff
@@ -331,6 +346,106 @@ public boolean isValueSetSupported(ValidationSupportContext theValidationSupport
...
+		IBaseParameters params = ParametersUtil.newInstance(fhirContext);
+
+		if (!StringUtils.isEmpty(theRequest.getConceptMapUrl())) {
+			ParametersUtil.addParameterToParametersUri(fhirContext, params, "url", theRequest.getConceptMapUrl());
+		}
+		if (!StringUtils.isEmpty(theRequest.getConceptMapVersion())) {
+			ParametersUtil.addParameterToParametersString(fhirContext, params, "conceptMapVersion", theRequest.getConceptMapVersion());
+		}
+		if (theRequest.getCodings() != null) {
+			BaseRuntimeElementCompositeDefinition<?> codeableConceptDef = (BaseRuntimeElementCompositeDefinition<?>) Objects.requireNonNull(fhirContext.getElementDefinition("CodeableConcept"));
```
Comment (fil512) [inline hapi-fhir-validation/src/main/java/org/hl7/fhir/common/hapi/validation/support/RemoteTerminologyServiceValidationSupport.java:362]: suggestion: TerserUtil might have helper methods in it that could simplify what you're doing here.

File: hapi-fhir-validation/src/test/java/org/hl7/fhir/common/hapi/validation/support/RemoteTerminologyServiceValidationSupportTest.java [hunk 3]
```diff
@@ -277,6 +297,91 @@ public void testValidateCodeInValueSet_InferSystem() {
 		assertEquals(null, outcome);
 	}
 
+	@Test
+	public void testTranslateCode_AllInParams_AllOutParams() {
+		myConceptMapProvider.myNextReturnParams = new Parameters();
```
Comment (fil512) [inline hapi-fhir-validation/src/test/java/org/hl7/fhir/common/hapi/validation/support/RemoteTerminologyServiceValidationSupportTest.java:302]: once this MR merges, I recommend adding a test to a cdr-app IT that uses a remote terminology service to ensure all the plumbing works as expected.  (Or at a minimum testing with cdr manually.)

**Review by fil512 — APPROVED** [review]

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/dstu3/FhirResourceDaoConceptMapDstu3.java [hunk 4]
```diff
@@ -38,19 +39,29 @@
...
 	private ITermConceptMappingSvc myTermConceptMappingSvc;
+	@Autowired
+	private IValidationSupport myValidationSupport;
 
 	@Override
 	public TranslateConceptResults translate(TranslationRequest theTranslationRequest, RequestDetails theRequestDetails) {
-		if (theTranslationRequest.hasReverse() && theTranslationRequest.getReverseAsBoolean()) {
-			return myTermConceptMappingSvc.translateWithReverse(theTranslationRequest);
-		}
+		IValidationSupport.TranslateCodeRequest translateCodeRequest = new IValidationSupport.TranslateCodeRequest(
```
[the remark this answers is not in the document]
Comment (seanmikalson-smilecdr) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/dstu3/FhirResourceDaoConceptMapDstu3.java]: Done!

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/TermConceptMappingSvcImpl.java (hunk 1 above)
Comment (seanmikalson-smilecdr) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/TermConceptMappingSvcImpl.java]: Done!

File: hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/term/TermConceptMappingSvcImplTest.java [hunk 5]
```diff
@@ -1547,6 +1559,104 @@ protected void doInTransactionWithoutResult(@Nonnull TransactionStatus theStatus
...
+		mock.translateConcept(theRequest);
+		verify(mock).translate(argument.capture());
+		assertTrue(expected.getCodeableConcept().equalsDeep(argument.getValue().getCodeableConcept()));
+		assertEquals(expected.getConceptMapVersion().asStringValue(), argument.getValue().getConceptMapVersion().asStringValue());
+		assertEquals(expected.getUrl().asStringValue(), argument.getValue().getUrl().asStringValue());
+		assertEquals(expected.getSource().asStringValue(), argument.getValue().getSource().asStringValue());
+		assertEquals(expected.getTarget().asStringValue(), argument.getValue().getTarget().asStringValue());
+		assertEquals(expected.getTargetSystem().asStringValue(), argument.getValue().getTargetSystem().asStringValue());
+		assertEquals(expected.getResourceId(), argument.getValue().getResourceId());
+		assertEquals(expected.getReverseAsBoolean(), argument.getValue().getReverseAsBoolean());
```
[the remark this answers is not in the document]
Comment (seanmikalson-smilecdr) [reply hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/term/TermConceptMappingSvcImplTest.java]: Done!

File: hapi-fhir-validation/src/main/java/org/hl7/fhir/common/hapi/validation/support/RemoteTerminologyServiceValidationSupport.java (hunk 2 above)
Comment (seanmikalson-smilecdr) [reply hapi-fhir-validation/src/main/java/org/hl7/fhir/common/hapi/validation/support/RemoteTerminologyServiceValidationSupport.java]: Checked TerserUtil but couldn't find anything that could work for just a CodeableConcept; most methods looked like they were working with IBaseResource and not just a CodeableConcept. I could be missing something.

File: hapi-fhir-validation/src/main/java/org/hl7/fhir/common/hapi/validation/support/RemoteTerminologyServiceValidationSupport.java [hunk 6]
```diff
@@ -331,6 +346,106 @@ public boolean isValueSetSupported(ValidationSupportContext theValidationSupport
...
+		if (!StringUtils.isEmpty(theRequest.getTargetValueSetUrl())) {
+			ParametersUtil.addParameterToParametersUri(fhirContext, params, "target", theRequest.getTargetValueSetUrl());
+		}
+		if (!StringUtils.isEmpty(theRequest.getTargetSystemUrl())) {
+			ParametersUtil.addParameterToParametersUri(fhirContext, params, "targetsystem", theRequest.getTargetSystemUrl());
+		}
+		if (theRequest.isReverse()) {
+			ParametersUtil.addParameterToParametersBoolean(fhirContext, params, "reverse", theRequest.isReverse());
+		}
+
```
[the remark this answers is not in the document]
Comment (seanmikalson-smilecdr) [reply hapi-fhir-validation/src/main/java/org/hl7/fhir/common/hapi/validation/support/RemoteTerminologyServiceValidationSupport.java:355]: Made smaller methods. There's some ~25 liners still. Let me know your thoughts!

File: hapi-fhir-validation/src/test/java/org/hl7/fhir/common/hapi/validation/support/RemoteTerminologyServiceValidationSupportTest.java [hunk 7]
```diff
@@ -277,6 +297,91 @@ public void testValidateCodeInValueSet_InferSystem() {
...
+			.setConceptMapUrl(CONCEPT_MAP_URL)
+			.setDisplay(TARGET_CODE_DISPLAY);
+		translateResults.add(singleResult);
+		translateResults.add(singleResult);
+		expectedResults.setResults(translateResults);
+
+		CodeableConcept codeableConcept = new CodeableConcept();
+		codeableConcept.addCoding(new Coding(CODE_SYSTEM, CODE, null));
+
+		IValidationSupport.TranslateCodeRequest request = new IValidationSupport.TranslateCodeRequest(
```
[the remark this answers is not in the document]
Comment (seanmikalson-smilecdr) [reply hapi-fhir-validation/src/test/java/org/hl7/fhir/common/hapi/validation/support/RemoteTerminologyServiceValidationSupportTest.java:328]: Left as is.

File: hapi-fhir-validation/src/test/java/org/hl7/fhir/common/hapi/validation/support/RemoteTerminologyServiceValidationSupportTest.java (hunk 3 above)
Comment (seanmikalson-smilecdr) [reply hapi-fhir-validation/src/test/java/org/hl7/fhir/common/hapi/validation/support/RemoteTerminologyServiceValidationSupportTest.java:302]: I tested with cdr manually as part of making the change but happy to add an IT as well.

**Pull request merged**